### ISSUE TITLE
How to deal with HeteroData objects in ReFine?

### ISSUE THREAD

**Issue by MajLuu** [Wuyxin/ReFine#14]
I encouter problems when I use [HAN model](https://github.com/pyg-team/pytorch_geometric/blob/66b17806b1f4a2008e8be766064d9ef9a883ff03/torch_geometric/nn/conv/han_conv.py) to do some graph classification tasks. I read mutag datasets codes which deal with datasets with torch_geometric.data.Data, but HeteroData is used in HAN model.
I want to know if ReFine can deal with HeteroData objects.

**Comment by Wuyxin** [comment]
Hi, 

It would be great if you have the specific code example that gives the error information etc. 

And by my understanding, your question seems to be more about the GNN model and does not relate to the explaining process. But I can see how I can help.

**Comment by MajLuu** [comment]
The example code as follows, this task is a graph classification problem.(Maybe can not run, but the meaning is right.)
```python
from typing import Dict, Union

import torch
import torch.nn.functional as F
from torch import nn
from torch_geometric.nn.conv import HANConv
from torch_geometric.data import HeteroData, DataLoader
from torch_geometric.nn import global_mean_pool


class HAN(nn.Module):
    def __init__(self, in_channels: Union[int, Dict[str, int]],
                 out_channels: int, hidden_channels=128, heads=8, metadata=None):
        super().__init__()
        self.han_conv = HANConv(in_channels, hidden_channels, heads=heads,
                                dropout=0.6, metadata=metadata)
        self.lin = nn.Linear(hidden_channels, out_channels)

    def forward(self, x_dict, edge_index_dict, batch):
        out = self.han_conv(x_dict, edge_index_dict)
        out = global_mean_pool(out['author'], batch=batch)
        return self.net(out)


def test_han_conv():
    data = HeteroData(y=torch.LongTensor([1]))  # 假设为二分类问题
    data['author'].x = torch.randn(6, 16)
    data.num_nodes = 6

    edge1 = torch.randint(0, 6, (2, 7), dtype=torch.long)
    edge2 = torch.randint(0, 5, (2, 4), dtype=torch.long)
    edge3 = torch.randint(0, 3, (2, 5), dtype=torch.long)
    data['author', 'metapath0', 'author'] = edge1
    data['author', 'matapath1', 'author'] = edge2
    data['author', 'matapath2', 'author'] = edge3

    in_channels = {'author': 16}
    model = HAN(in_channels=in_channels, out_channels=1, heads=8, metadata=data.metadata())
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    data, model = data.to(device), model.to(device)
    optimizer = torch.optim.Adam(model.parameters(), lr=0.005, weight_decay=0.001)
    train_loader = DataLoader(data, batch_size=1, shuffle=False)

    def train(train_loader) -> float:
        model.train()
        for data in train_loader:
            data.to(device)
            optimizer.zero_grad()
            out = model(data.x_dict,
                        data.edge_index_dict,
                        data['author'].batch)
            loss = F.cross_entropy(out, data.y)
            loss.backward()
            return float(loss)

    for epoch in range(1, 200):
        loss = train(train_loader)
        torch.save(model.cpu(), "gnn.pt")   # 训练一次做个样子，假设得到最终模型
        break
```
The example code uses HeteroData, and edge list is not sequential which can not be used edge_attr to identify one edge belonging to which metapath. If i want to explain "gnn.pt", how I set ReFine model parameters? Or how I modify example code which will be suitable for ReFine?
```python
explainer = ReFine(device, "gnn.pt", gamma=args.gamma,
                       n_in_channels=torch.flatten(data['author'].x, 1, -1).size(1),
                       # e_in_channels怎么配置？
                       e_in_channels=train_dataset[0].edge_attr.size(1),
                       n_label=2)
```

**Comment by Wuyxin** [comment]
I see. It may take some work to adapt in this case as ReFine is built upon `pyg.data.Data`, and the edge mask setting in pyg could only be set once before propagating while HANConv propagates the message multiple times.

If you are to do so, you may need to (1) set the edge mask multiple times with the loops containing self.propogate in HANConv  (2) modify all input of self.model() in ReFine (3) adapt the data type for generating edge masks, e.g.,
```
data_homo = data.to_homogeneous # to pyg.data.Data type
data_homo.edge_attr = torch.ones((data_homo.num_edges, 1))
# then input data_homo to the EdgeMask Net, and get edge mask 
# set edge mask for Han
# adapt self.model() in ReFine to align with your forward args (x_dict, edge_index_dict, batch)
```
